agent/utils/utils_windows.go

```diff
@@ -0,0 +1,81 @@
+// +build windows
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package utils
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+func GetCanonicalPath(path string) string {
+	lowercasedPath := strings.ToLower(path)
+	// if the path is a bare drive like "d:", don't filepath.Clean it because it will add a '.'.
+	// this is to fix the case where mounting from D:\ to D: is supported by docker but not ecs
+	if isBareDrive(lowercasedPath) {
+		return lowercasedPath
+	}
+
+	if isNamedPipesPath(lowercasedPath) {
+		return lowercasedPath
+	}
+
+	return filepath.Clean(lowercasedPath)
+}
+
+func isBareDrive(path string) bool {
+	if filepath.VolumeName(path) == path {
+		return true
+	}
+
+	return false
+}
+
+func isNamedPipesPath(path string) bool {
+	matched, err := regexp.MatchString(`\\{2}\.[\\]pipe[\\].+`, path)
+
+	if err != nil {
+		return false
+	}
+
+	return matched
+}
+
+// findUnusedDriveLetter is used to search for an available drive letter on the container instance.
+// Reference: https://golang.org/src/os/os_windows_test.go
+func FindUnusedDriveLetter() (string, error) {
+	// Do not use A: and B:, because they are reserved for floppy drive.
+	// Do not use C:, because it is normally used for main drive.
```

Comment:
You are making 22 system calls (checking driver letters from Z to D ) for every volume , which is a costly operation, Is it not possible to optimize ?   Had we considered the approach of storing the drive letters in use in memory or writing to the ecs_agent json file used to preserve the state , that way you will be able to findout a driver letter available just in one call  which saves a lot of compute ?

Reply:
what happens if the instance owner or user uses some drive letter that we assume to be available? static checks are not a feasible approach, hence we need to do this check

Reply:
Ok , agreed . In that way can't we get all the used drive letters in one system call  and then use one drive letter which is not used ? I am just trying to see how we can avoid too many calls here  and i feel there must be an api to get all used drive letters on windows (like this one powershell Get-PSDrive -PSProvider FileSystem)  , please take a look